subsys/debug/asan.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2019 Jan Van Winkel <jan.van_winkel@dxplore.eu>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_SOC_POSIX)
+#include <sanitizer/lsan_interface.h>
+#include "soc.h"
+#endif
+
+const char *__asan_default_options()
+{
+	return
+#ifdef CONFIG_TEST
+		"log_path=stdout:"
+#endif
+#if defined(CONFIG_64BIT) && defined(__GNUC__) && !defined(__clang__)
+		/* Running leak detection at exit could lead to a deadlock on
+		 * 64-bit boards if GCC is used.
+		 * https://github.com/zephyrproject-rtos/zephyr/issues/20122
+		 */
+		"leak_check_at_exit=0:"
```

Comment:
If we are setting this one, why do we need to also call `__lsan_do_leak_check()`?

Reply:
This flag prevents to run the leak checking `at exit` time in the global destructor which can lead to a deadlock. Note that leak detection bookkeeping logic is not disabled by the flag `leak_check_at_exit` and leak detection reporting can still be triggered via `__lsan_do_leak_check()`.

Further if `__lsan_do_leak_check()` is executed `at exit` leak detection is skipped and it could be argued that setting the default flag is not needed but this could lead to a deadlock if for some reason `__lsan_do_leak_check()` is not executed.

Reply:
Ok.. so you want to keep doing the leak check but in a kinda bad place...